File: Data-Structures/Stack/Stack.cs
namespace Stack;

public class Node
{
    public int Value { get; set; }
    public Node? Prev { get; set; }
}
public class StackList
{
    public Node? Head { get; set; }
    public int Size { get; set; }
    //print stack
    public void Print()
    {
        if(Head==null)
        {
            return;
        }
        Node curr = Head;
        while(curr != null)
        {
            Console.WriteLine(curr.Value);
            if(curr.Prev==null)return;
            curr=curr.Prev;
        }
    }
    //see first value of stack
    public Object? Peek()
    {
        return Head?.Value;
    }
    // add value to stack
    public void Push(int value)
    {   
        Node node = new Node
        {
            Value = value,
            Prev = Head
        };
        Head=node;
        Size++;
    }
    //remove and return fist value
    public Object? Pop()
    {
        if(Head==null)
        {
            return null;
        }
        Node curr = Head;
        Head = Head.Prev;
        return curr;
    }
}

File: Data-Structures/Stack/stack.go
package main

import "fmt"

type Node struct {
	Value int
	Next  *Node
}
type Stack struct {
	Top *Node
}

func Push(s *Stack, value int) {
	node := &Node{
		Value: value,
		Next:  s.Top,
	}
	s.Top = node
}

func Pop(s *Stack) (int, error) {

	if s.Top == nil {
		return 0, fmt.Errorf("Stack is empty")
	}
	value := s.Top.Value
	s.Top = s.Top.Next

	return value, nil
}

func Peek(s *Stack) (int, error) {
	if s.Top == nil {
		return 0, fmt.Errorf("Stack is empty")
	}
	return s.Top.Value, nil
}

func main() {
	mystack := Stack{}

	// Example usage:
	Push(&mystack, 42)
	Push(&mystack, 15)

	topValue, err := Peek(&mystack)
	if err != nil {
		fmt.Println(err)
	} else {
		fmt.Println("Top of the stack:", topValue)
	}

	poppedValue, err := Pop(&mystack)
	if err != nil {
		fmt.Println(err)
	} else {
		fmt.Println("Popped value:", poppedValue)
	}

	fmt.Println("Is the stack empty?", mystack.Top == nil)

}
